research/picai/preprocessing.py

```diff
@@ -0,0 +1,358 @@
+        """
+        case_origin, case_direction, case_spacing = None, None, None
+        for scan in mri.scans:
+            # copy metadata of first scan (nnUNet and nnDetection require this to match exactly)
+            if case_origin is None:
+                case_origin = scan.GetOrigin()
+                case_direction = scan.GetDirection()
+                case_spacing = scan.GetSpacing()
+            else:
+                # check if current scan's metadata is almost equal to the first scan
+                np.testing.assert_allclose(scan.GetOrigin(), case_origin)
+                np.testing.assert_allclose(scan.GetDirection(), case_direction)
+                np.testing.assert_allclose(scan.GetSpacing(), case_spacing)
+
+                # copy over first scan's metadata to current scan
+                scan.SetOrigin(case_origin)
+                scan.SetDirection(case_direction)
+                scan.SetSpacing(case_spacing)
+
+        assert case_origin is not None and case_direction is not None and case_spacing is not None
+        mri.annotation.SetOrigin(case_origin)
+        mri.annotation.SetDirection(case_direction)
+        mri.annotation.SetSpacing(case_spacing)
+
+        return mri
+
+
+class BinarizeAnnotation(PreprocessingTransform):
+    def __call__(self, mri: Case) -> Case:
+        """
+        Binarize the granular ISUP ≥ 2 annotations.
+
+        Args:
+            mri (Case): The case to be processed.
+
+        Returns:
+            Case: The Case after binarizing the annotation.
+        """
+        annotation_array = sitk.GetArrayFromImage(mri.annotation)
+
+        # convert granular PI-CAI csPCa annotation to binary csPCa annotation
+        annotation_array = (annotation_array >= 1).astype("uint8")
+
+        # convert label back to SimpleITK
+        new_annotation = sitk.GetImageFromArray(annotation_array)
+        new_annotation.CopyInformation(mri.annotation)
+        mri.annotation = new_annotation
+        return mri
+
+
+def apply_transform(mri: Case, transforms: Sequence[PreprocessingTransform]) -> Tuple[Sequence[Path], Path]:
+    """
+    Reads in scans and annotation, applies sequence of transformations, and writes resulting case to disk.
+    Returns tuple with scan paths and corresponding annotation path.
+
+    Args:
+        mri (Case): The case to be processed.
+        transforms (Sequence[PreprocessingTransform]): The sequence of transformation to be applied.
+
+    Returns:
+        Tuple[Sequence[Path], Path]: A tuple in which the first entry is a sequence of file paths
+            for the scans and the second entry is the file path to the correpsonding annotation.
+
+    Raises:
+        PreprocessingException if an error occurs during preprocessing.
+    """
+    try:
+        mri.read()
+        processed_mri = reduce(lambda acc, f: f(acc), transforms, mri)
+        return processed_mri.write()
+    except Exception as e:
```

Comment:
Sometimes we get a flake (or some other linter) complaint about a bare exception here. It seems like you're passing those checks for now, but just wanted to note that you can ignore it if that pops up.

Reply:
Okay I will note that! The only reason I catch the bare exception is so I can also surface what caused the PreProcessingException. If there is a better way to do this definitely, let me know :)

Reply:
Yeah I figured that's why you're doing that. I'm perfectly fine with that. Just wanted to note that sometimes linters complain about this 😂